### ISSUE TITLE
Liberalize handling of facts gathering

### ISSUE THREAD

**Issue by jeffbrl** [Juniper/py-junos-eznc#461]
I believe facts gathering should not fail when some facts are missing. In my 14.1R1.10 vMX router, 'show chassis inventory' does not emit a serial number for the chassis (among other problems). 

Shouldn't the more pythonic way of executing the functions in FACT_LIST be more forgiving? We'll have to keep fixing the functions for platform variations; however, I think the script should continue with the facts it has gathered.

From device.py:

```
       for gather in FACT_LIST:
            try:
                gather(self, self._facts)
            except:
                if exception_on_failure:
                    raise
                warnings.warn('Facts gathering is incomplete. '
                              'To know the reason call "dev.facts_refresh(exception_on_failure=True)"', Runtim
eWarning)
                return
```

I'd be glad to help out on a fix if I get guidance from the core developers.

**Comment by stacywsmith** [comment]
@jeffbrl 
That's essentially what the recent commit in #459 should do.
A new parameter, exception_on_failure, was added to the facts_refresh() method, and exceptions will only be raised if that parameter is True.

By default, the parameter is false and exceptions are not raised.
def facts_refresh(self, exception_on_failure=False):

If you still prefer the previous behavior, you can set exception_on_failure=True.

Are you testing with the latest PyEZ from GitHub?

Is there something that #459 does not address to your satisfaction?

**Comment by jeffbrl** [comment]
@stacywsmith I pip installed from master today, so I have the fix for #459. 

Let me re-examine this issue as vMX-specific one. If it is, we can close this issue and I'll open a new issue.

**Comment by jeffbrl** [comment]
I got confused when an IndexError exception raised by an ansible-generated python script appeared immediately after the facts gathering warning. Unfortunately, I removed the temporary python script and can't get my playbook to trigger it again.

For reference purposes, the only facts for my vMX version are {'2RE': False, 'model': 'VMX'}.

Let's chalk this one up to operator error for now. I'll open an issue with Juniper.junos ansible module if there is a repeatable problem with my version of vMX.

Thanks, @stacywsmith.
